### PR TITLE
Add custom mailservers

### PR DESCRIPTION
**Pull request by cammellos**
### Summary:

This is the first step of adding mailserver:

It allows to add a custom mailserver by entering the url. You can then connect to it and use. Mailservers are preserved, partially addressing https://github.com/status-im/status-react/issues/4328.

I kept the changes similar to the recently merged PR of managing networks for consistency, with the difference that mailservers are stored in the account realm rather than base.

Next iteration would be to add QR code scanning & deleting mailservers.

### Testing notes (optional):

If upgrading from a previous version it will reset the database, I was asked to leave it like so until we cut the next release.

There's another bug that you might come across:
`web3.shh null is not an object`.

This happens on logout when the connection to the mailserver is not established, at which point the connection might succeed. This has not been address as there's already some code in the testing pipeline which will change the behaviour. I could reproduce in develop.

The feature is now hidden behind a flag, by default is shown but I have toggled it off for nightly & production, pending a discussion in slack.

status: ready

### CODE REVIEW

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/7zyCnt for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/7zyCnt">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/iyhxe5 for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/iyhxe5">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

File: src/status_im/data_store/core.cljs [hunk 1]
```diff
@@ -19,9 +20,16 @@
 (defn change-account [address new-account? encryption-key handler]
   (data-source/change-account address new-account? encryption-key handler))
 
-(defn- perform-transactions [transactions realm]
-  (data-source/write realm #(doseq [transaction transactions]
-                              (transaction realm))))
+(defn- perform-transactions [raw-transactions realm]
```
Comment (yenda) [inline src/status_im/data_store/core.cljs:23]: could you specify the purpose of this change ?

File: src/status_im/data_store/realm/core.cljs [hunk 2]
```diff
@@ -70,12 +70,7 @@
   ;; delete the realm file if its schema version is higher
   ;; than existing schema version (this means the previous
   ;; install has incompatible database schemas)
-  (if-let [current-version (realm-version file-name encryption-key)]
-    (if (> current-version
-           (apply max (map :schemaVersion base/schemas)))
-      (reset-realm file-name schemas encryption-key)
-      (migrate-realm file-name schemas encryption-key))
-    (reset-realm file-name schemas encryption-key)))
+  (migrate-realm file-name schemas encryption-key))
```
Comment (yenda) [inline src/status_im/data_store/realm/core.cljs:64]: could this be done in a separate commit that we merge right after 0.19 cut ?

File: src/status_im/transport/inbox.cljs [hunk 3]
```diff
@@ -179,6 +180,13 @@
 (def ^:private ^:const max-retries 10)
 (def ^:private ^:const retries-interval-change-threshold 3)
 
+(defn add-custom-mailservers [mailservers {:keys [db]}]
```
Comment (yenda) [inline src/status_im/transport/inbox.cljs:183]: I think we should do things differently and consider your new get-all-mailserver cofx to be the source of all mailservers, so when we create an account we save the default ones and there is no need to merge them everytime we login

File: src/status_im/data_store/core.cljs (hunk 1 above)
Comment (cammellos) [reply src/status_im/data_store/core.cljs:23]: I was tracking down a bug where by it would try to write to a closed realm, turned out it was a different issue, but the problem still stand that writes to realm are asynchronous (they go through a core async queue), so some actions do depend on it to be completed, so we should only `logout` once the realm db transaction has been successfully written.

File: src/status_im/data_store/realm/core.cljs (hunk 2 above)
Comment (cammellos) [reply src/status_im/data_store/realm/core.cljs:64]: Yes, no probs, it will try to reset the db, but as long as we are aware of it I see no problems

File: src/status_im/transport/inbox.cljs (hunk 3 above)
Comment (cammellos) [reply src/status_im/transport/inbox.cljs:183]: @yenda you mean storing the default mailservers in realm as well? 

I thought about that, but I decided against it, as it would make a bit harder changing them ( we would have to migrate realm data). Also you would have some code that stores them once on account creation, so it would not just be a matter of removing code and simplifying.

Personally it seems simpler to just merge the custom one on login, rather than making the default one persisted in realm, less stateful. 

What do you think?

File: src/status_im/ui/screens/events.cljs [hunk 4]
```diff
@@ -263,6 +263,7 @@
                                               [:load-accounts]
                                               [:listen-to-network-status]
                                               [:navigate-to :accounts]]}
+                                (navigation/navigate-to-clean nil)
```
Comment (janherich) [inline src/status_im/ui/screens/events.cljs:266]: What will navigating to `nil` achieve ?

**Review by janherich — APPROVED** [review]

File: src/status_im/ui/screens/events.cljs (hunk 4 above)
Comment (cammellos) [reply src/status_im/ui/screens/events.cljs:266]: @janherich we go back to the initial state and removes the subscription (database starts with `nil` in `view-id`). 

Basically if we don't subscriptions as still fired on database changes, but `account/account` (for example) is not there. 

In this example after `re-frame` process `:initialize-db` it will try to render the current `:view-id` but has not the correct data. Does it make sense?

File: src/status_im/data_store/realm/schemas/account/v2/mailserver.cljs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+(ns status-im.data-store.realm.schemas.account.v2.mailserver
+  (:require [taoensso.timbre :as log]))
+
+(def schema {:name       :mailserver
+             :primaryKey :id
+             :properties {:id        :string
+                          :name      {:type     :string}
+                          :address   {:type     :string}
+                          :password  {:type   :string
+                                      :optional true}
```
Comment (dmitryn) [inline src/status_im/data_store/realm/schemas/account/v2/mailserver.cljs]: map values align

File: src/status_im/transport/inbox.cljs [hunk 6]
```diff
@@ -179,6 +180,13 @@
...
+  {:db (reduce (fn [db {:keys [id chain] :as mailserver}]
+                 (assoc-in db [:inbox/wnodes (keyword chain) id]
+                           (-> mailserver
+                               (dissoc :chain)
+                               (assoc :user-defined true)))) db mailservers)})
```
Comment (dmitryn) [inline src/status_im/transport/inbox.cljs]: I would prefer `reduce` arguments to be vertically align https://github.com/bbatsov/clojure-style-guide#vertically-align-fn-args
```clj
(reduce (fn [] ...
        db
        mailservers)
```

**Review by dmitryn — APPROVED** [review]
LGTM, added some style comments

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/L2bAw3 for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/L2bAw3">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/9aAF8A for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/9aAF8A">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/null for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/null">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/hswgKt for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/hswgKt">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/E9EHQL for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/E9EHQL">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/2qvfEw for branch PR-4372

<blockquote><div><strong><a href="https://i.diawi.com/2qvfEw">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>
